trade-finance-manager-ui/server/controllers/utilisation-reports/index.ts

```diff
@@ -1,30 +1,130 @@
     console.error('Error rendering utilisation reports page', error);
     return res.render('_partials/problem-with-service.njk', { user });
   }
 };
+
+const getReportIdentifiersFromUrl = (body: object, month: number, year: number): ReportIdentifier[] =>
+  Object.keys(body).reduce((reportIdentifiers, key) => {
+    const match = key.match(/(set-status--)(.*)(-)([a-f\d]+)/);
+    if (!match) {
+      return reportIdentifiers;
+    }
+
+    const idType = match.at(2);
+    const id = match.at(4);
+
+    if (!idType || !id) {
+      throw new Error('Failed to parse report identifiers from request body');
+    }
+
+    if (idType === 'reportId') {
+      return [...reportIdentifiers, { id }];
+    }
+
+    if (idType === 'bankId') {
+      return [...reportIdentifiers, { bankId: id, month, year }];
+    }
+
+    throw new Error('Failed to parse report identifiers from request body');
+  }, [] as ReportIdentifier[]);
+
+export type UpdateReportStatusPayload = {
+  reportsWithStatus: ReportWithStatus[];
+  user: TfmSessionUser;
+};
+
+export const updateUtilisationReportStatus = async (req: Request, res: Response) => {
+  const { user, userToken } = req.session;
+  const { status, month, year } = req.query;
+
+  try {
+    if (!user) {
+      throw new Error('User is not logged in');
+    }
+
+    const reportIdentifiers = getReportIdentifiersFromUrl(req.body as object, Number(month), Number(year));
+
+    if (status !== 'completed' && status !== 'not-completed') {
+      throw new Error('Supplied status instruction does not match any expected status');
+    }
+    const statusToSet =
+      status === 'completed' ? UTILISATION_REPORT_RECONCILIATION_STATUS.RECONCILIATION_COMPLETED : UTILISATION_REPORT_RECONCILIATION_STATUS.REPORT_NOT_RECEIVED;
+
+    const reportsWithStatus = reportIdentifiers.map((reportIdentifier): ReportWithStatus => {
+      if ('id' in reportIdentifier && statusToSet === UTILISATION_REPORT_RECONCILIATION_STATUS.REPORT_NOT_RECEIVED) {
+        return {
+          status: UTILISATION_REPORT_RECONCILIATION_STATUS.PENDING_RECONCILIATION,
+          report: reportIdentifier,
+        };
+      }
+      if ('bankId' in reportIdentifier && statusToSet === UTILISATION_REPORT_RECONCILIATION_STATUS.REPORT_NOT_RECEIVED) {
+        throw new InvalidUpdateReportStatusRequestError("Cannot set non-existing report to 'REPORT_NOT_RECEIVED'");
+      }
+      return { status: statusToSet, report: reportIdentifier };
+    });
+
+    const updateReportStatusPayload: UpdateReportStatusPayload = { user, reportsWithStatus };
+    await api.updateUtilisationReportStatus(updateReportStatusPayload, asString(userToken));
+
+    return await getUtilisationReports(req, res);
+  } catch (error) {
+    console.error('Error updating utilisation report statuses:', error);
+    if (error instanceof InvalidUpdateReportStatusRequestError) {
+      res.locals.errorSummary = [{ text: 'Cannot mark "not received" report as "not done"', href: '#' }];
```

Comment:
`text` - we should use '"not completed"` instead of `"not done"` to match the text of the button. If we're going to use a custom error for this case we could just use the message (or some other custom field if we choose) for the text here.

`href` - I don't think we want an anchor link for this error? In which case we should remove this attribute (see [docs](https://design-system.service.gov.uk/components/error-summary/#options-error-summary-example--error-list))

Reply:
Removed as there is no specific design for this error case, but as far as my testing showed an `href` is required in order to render the `govukErrorSummary` component